File: Go/410. Split Array Largest Sum.go
package main

//goland:noinspection GoUnusedFunction
func splitArray(nums []int, k int) int {
	sum := 0
	maximum := 0

	for _, num := range nums {
		sum += num
		maximum = max(maximum, num)
	}

	return binarySearch(nums, k, sum, maximum)
}

func binarySearch(nums []int, k int, high int, low int) int {
	for low <= high {
		mid := low + (high-low)/2
		if validSearch(nums, k, mid) {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	return low
}

func validSearch(nums []int, k int, sum int) bool {
	total := 0
	count := 1

	for _, num := range nums {
		total += num
		if total > sum {
			total = num
			count++
			if count > k {
				return false
			}
		}
	}

	return true
}

File: PHP/410. Split Array Largest Sum.php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    /**
     * @param  int[]  $nums
     */
    public function splitArray(array $nums, int $k): int
    {
        $sum = array_sum($nums);
        $max = max($nums);

        return $this->binarySearch($nums, $k, $sum, $max);

    }

    private function binarySearch(array $nums, int $k, int $high, int $low): int
    {

        while ($low <= $high) {
            $mid = (int) ($low + ($high - $low) / 2);
            if ($this->valid($nums, $k, $mid)) {
                $high = $mid - 1;
            } else {
                $low = $mid + 1;
            }
        }

        return $low;
    }

    private function valid(array $nums, int $k, int $sum): bool
    {
        $total = 0;
        $count = 1;

        foreach ($nums as $num) {
            $total += $num;
            if ($total > $sum) {
                $total = $num;
                $count++;

                if ($count > $k) {
                    return false;
                }
            }
        }

        return true;
    }
}
